tests/unit/lms/security_test.py

```diff
@@ -161,102 +167,82 @@ def test_authenticated_userid(self, lti_user, expected_userid, pyramid_request):
 
 
 class TestSecurityPolicy:
-    def test_it_returns_the_userid_from_LTISecurityPolicy(
-        self, policy, LTISecurityPolicy, lti_security_policy
-    ):
-        lti_security_policy.authenticated_userid.return_value = "user"
-        user_id = policy.authenticated_userid(mock.sentinel.request)
+    def test_authenticated_userid(self, policy, get_route_policy):
```

Comment:
For the methods that just call the right method on the chosen policy I'm just asserting that. Not any of the specifics of the configuration.

Reply:
I think it's a bit of a code smell for the tests for a class to mock out part of that class. We want cleanly defined public interfaces between the different units of code that we're testing. Treating `SecurityPolicy.get_route_policy()` as one unit and the rest of the `SecurityPolicy` class as a separate unit doesn't seem very clearly defined to me since they're two parts of the same class. What other parts of the `SecurityPolicy` class are used by `get_route_policy()`? And which by the other methods? And which by both? In fact `SecurityPolicy.path_policy_mapping` and `SecurityPolicy.fallback_policy` are only used by `SecurityPolicy.get_route_policy()`, and the rest of `SecurityPolicy`'s methods only use `get_route_policy()`, but you have to read the code of the whole `SecurityPolicy` class to spot that division.

I wonder if it might be clearer if that division were made explicit by splitting it into two separate classes?

Here's a rough sketch:

```python
class PathPolicyGetter:
    """Callable that returns different policies based on the request path."""

    def __init__(self, path_policy_mapping, fallback_policy):
        ...

    @lru_cache
    def __call__(self, request):
        # Return the matching policy from path_policy_mapping or fallback_policy.

class SecurityPolicy:
    def __init__(self, policy_getter):
        self._policy_getter = policy_getter

    def authenticated_userid(self, request):
        return self._policy_getter(request).authenticated_userid(request)

    ...

def includeme(config):
    policy_getter = PathPolicyGetter(
        path_policy_mapping={...},
        fallback_policy=LTISecurityPolicy
    )
    config.set_security_policy(SecurityPolicy(PathPolicyGetter))
```

Then you can unit test `PathPolicyGetter` which has no dependencies (other than the injected mapping dict). And you can separately test `SecurityPolicy` by passing it in a mock `policy_getter`. You won't need to patch anything

Reply:
Went with pretty much that solution verbatim.

I started with something very similar but with an extra layer of subclassing  (eg `PolicyGetter -> `PathPolicyGetter`) but I was adding to much "generalization" and scratched it. Might have gone to far in the other direction. 

I don't think we ever use more than one `PolicyGetter` at the same time. This makes the testing more convenient as you point out.